### DISCUSSION TITLE
How to write and read from collections

### DISCUSSION THREAD

**Discussion by alexander-schranz** [solariumphp/solarium#1049] [Q&A] [answered]
I'm trying to index documents in newly created collections. I know that I can define a specific connection at create a client, but I need to read and write into multiple collection but I'm not sure how to achieve this.

```php
<?php

use Solarium\Client;
use Solarium\Core\Client\Adapter\Curl;

// create Client
$adapter = new Curl();
$eventDispatcher = new EventDispatcher();
$options = [
    'endpoint' => array(
        'localhost' => array(
            'host' => $host,
            'port' => $port,
        ),
    )
];

$client = new Client($adapter, $eventDispatcher, $options);

// create collection test_simple
$collectionQuery = $this->client->createCollections();

$action = $collectionQuery->createDelete(['name' => 'test_simple']);
$collectionQuery->setAction($action);

$this->client->collections($collectionQuery);

// create collection test_complex
$collectionQuery = $this->client->createCollections();

$action = $collectionQuery->createDelete(['name' => 'test_complex']);
$collectionQuery->setAction($action);

$this->client->collections($collectionQuery);

// now I try to index different documents into different collections but it seems to fail:
$update = $this->client->createUpdate([
    'collection' => 'test_simple',
]);


$indexDocument = $update->createDocument(['id' => 'c1ee078b-289e-42e6-8040-d5ea3288e804', 'title' => 'Test']);

$update->addDocuments([$indexDocument], true);

$update->addCommit();

$result = $this->client->update($update);
```

But it fails with:

> Solarium\Exception\HttpException: Solr HTTP error: Neither collection nor core set. (404)

I'm setting the collectino as `option` so I'm not sure how to achieve this corectly?

**Comment by thomascorthals** [comment]
You're calling the wrong method to create collection (twice). This should be `createCreate()` instead of `createDelete()`.

```php
$action = $collectionQuery->createDelete(['name' => 'test_simple']);
```

After that, you have to add an endpoint that points to your new collection to the client instance.

```php
$endpoint = [
    'key' => 'test_simple', // this is how you want to identify the endpoint, doesn't have to be the collection name
    'host' => $host,
    'port' => $port,
    'collection' => 'test_simple',
];

$this->client->addEndpoint($endpoint);
```

You can now use the endpoint key when executing an update (or any other) query.

```php
$update = $this->client->createUpdate(); // no options needed here for your use case

// add documents, commit

$result = $this->client->update($update, 'test_simple'); // execute against endpoint identified as 'test_simple'
```

**Reply by alexander-schranz** [reply]
Oh that was a typo. The hint with the collection did work. Really thank you for your fast reponse. But as the collection is in my case dynamic and I and the collections are not known when the client is created I did  go with:

```php
$this->client->getEndpoint()
    ->setCollection('test_simple');
```

**Comment by thomascorthals** [comment, marked as answer]
From your question regarding schemas, I assume you're trying to configure which schema to use for your new collection? You can specify the name of a configuration to use for the collection.

```php
$action = $collectionQuery->createCreate();
$action->setName('test_simple')
       ->setCollectionConfigName('techproducts');
```

(I prefer the above syntax, but `$collectionQuery->createCreate(['name' => 'test_simple', 'collection.configName' => 'techproducts'])` also works.)

If you're not sure what "name of a configuration" means here and how it relates to the schema, start with reading [configsets](https://solr.apache.org/guide/solr/latest/configuration-guide/config-sets.html) in the ref guide.

**Reply by alexander-schranz** [reply]
In our case we need to create the Schema/FieldMapping programmatically as it is defined by the Application. So the schema cofnig really does not exist yet in Apache Solr Server.

That is why I did have a look at the [SchemaApi](https://solr.apache.org/guide/solr/latest/indexing-guide/schema-api.html). But that seems not be implemented in the Client.

And in the closed PR it did sound like there is another way to maybe push a schema file to the Apache Solr, but I could not yet find out how? Or is this something currently not supported?

**Reply by thomascorthals** [reply]
You can use the [Configsets API](https://solr.apache.org/guide/solr/latest/configuration-guide/configsets-api.html) to upload a configset to Zookeeper.

https://github.com/solariumphp/solarium/blob/3e0eb01c8ef02554b5cc5755019704cce6b12d02/examples/execute_all.php#L37-L44

An existing schema can be updated through the [Schema API](https://solr.apache.org/guide/solr/latest/indexing-guide/schema-api.html).

https://github.com/solariumphp/solarium/blob/3e0eb01c8ef02554b5cc5755019704cce6b12d02/examples/execute_all.php#L101-L115

**Reply by thomascorthals** [reply]
@alexander-schranz I've updated my comment above with example code to upload a configset. Does this answer your question?

**Reply by alexander-schranz** [reply]
Thx for the response 👍, This looks what I was searching for. I think it should also work for nested objects (docs) but think that should be able to findout in the schema api docs.

Will give it a try this week, and give you an update here then. Thank you again for your help!

**Reply by alexander-schranz** [reply]
@thomascorthals Thank you your mention method over the configset and raw createApi did work for me. For later stumble over same problem the code can be found here: https://github.com/schranz-search/schranz-search/pull/73/files#diff-c873e4216e21cddd561c38583e6f9bdfe50253a8b1c9c9d9ffd5ac1ffcd39352 in the `createIndex` method. I just needed to add `SOLR_OPTS: '-Dsolr.disableConfigSetsCreateAuthChecks=true'` to my docker container else I was not able to create configssets

Next step for me will be to findout how [nested docs / objects](https://github.com/schranz-search/schranz-search/blob/1154fccb88afed9006255258a46e2305fb3ac7d1/packages/seal/Tests/Marshaller/MarshallerTest.php#L69-L90) works in solr over the schema api.

**Reply by alexander-schranz** [reply]
@thomascorthals Thx for your support. I did flatten all the objects now like elasticsearch is doing it internal, just on the PHP side to get nested json work also.

So I was able to merge: https://github.com/schranz-search/schranz-search/pull/73

And so the adapter is now available here: https://github.com/schranz-search/seal-solr-adapter
